### ISSUE TITLE
data_parallel_tree_learner error while installing GPU version of LightGBM with CUDA 10.2

### ISSUE THREAD

**Issue by Garrus990** [microsoft/LightGBM#3040]
First of all, thank you for the great work that you do! I appreciate it.

I have, however, a problem with installation of the GPU version of the package in a docker image where CUDA is already installed. I want to create a personalized docker with LightGBM with slightly newer version of CUDA than in the example dockerfile that is enclosed in the repository (where CUDA 8.0 is used). My starting point is then the image `nvidia/cuda:10.2-cudnn7-runtime-ubuntu18.04`. Everything goes smoothly until the moment when the following command is executed: `make OPENCL_HEADERS=/usr/local/cuda-10.2/targets/x86_64-linux/include LIBOPENCL=/usr/local/cuda-10.2/targets/x86_64-linux/lib`, which follows the lines (copied from the GPU dockerfile in the repository):
```
cd /usr/local/src && mkdir lightgbm && cd lightgbm && \
   git clone --recursive --branch stable --depth 1 https://github.com/microsoft/LightGBM && \
   cd LightGBM && mkdir build && cd build && \
   cmake -DUSE_GPU=1 -DOpenCL_LIBRARY=/usr/local/cuda/lib64/libOpenCL.so \
         -DOpenCL_INCLUDE_DIR=/usr/local/cuda/include/ ..
```
Let me underline, that the error doesn't occur, when the built is run as in the dockerfile in the repository (CUDA 8.0) - it finishes successfully.

## Environment info

Operating System: Ubuntu 18.04

CPU/GPU model: CUDA 10.2, CUDNN 7

C++/Python/R version: C++ 7.5.0 / Python 3.7.7

LightGBM version or commit hash: 471d0aa (stable)

## Error message
```
[root@bb7fa01ae535 build][(no branch)]# make OPENCL_HEADERS=/usr/local/cuda-10.2/targets/x86_64-linux/include LIBOPENCL=/usr/local/cuda-10.2/targets/x86_64-linux/lib
Scanning dependencies of target lightgbm
[  1%] Building CXX object CMakeFiles/lightgbm.dir/src/main.cpp.o
[  3%] Building CXX object CMakeFiles/lightgbm.dir/src/application/application.cpp.o
[  4%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/boosting.cpp.o
[  6%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/gbdt.cpp.o
[  7%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/gbdt_model_text.cpp.o
[  9%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/gbdt_prediction.cpp.o
[ 11%] Building CXX object CMakeFiles/lightgbm.dir/src/boosting/prediction_early_stop.cpp.o
[ 12%] Building CXX object CMakeFiles/lightgbm.dir/src/io/bin.cpp.o
[ 14%] Building CXX object CMakeFiles/lightgbm.dir/src/io/config.cpp.o
[ 15%] Building CXX object CMakeFiles/lightgbm.dir/src/io/config_auto.cpp.o
[ 17%] Building CXX object CMakeFiles/lightgbm.dir/src/io/dataset.cpp.o
[ 19%] Building CXX object CMakeFiles/lightgbm.dir/src/io/dataset_loader.cpp.o
[ 20%] Building CXX object CMakeFiles/lightgbm.dir/src/io/file_io.cpp.o
[ 22%] Building CXX object CMakeFiles/lightgbm.dir/src/io/json11.cpp.o
[ 23%] Building CXX object CMakeFiles/lightgbm.dir/src/io/metadata.cpp.o
[ 25%] Building CXX object CMakeFiles/lightgbm.dir/src/io/parser.cpp.o
[ 26%] Building CXX object CMakeFiles/lightgbm.dir/src/io/tree.cpp.o
[ 28%] Building CXX object CMakeFiles/lightgbm.dir/src/metric/dcg_calculator.cpp.o
[ 30%] Building CXX object CMakeFiles/lightgbm.dir/src/metric/metric.cpp.o
[ 31%] Building CXX object CMakeFiles/lightgbm.dir/src/objective/objective_function.cpp.o
[ 33%] Building CXX object CMakeFiles/lightgbm.dir/src/network/linker_topo.cpp.o
[ 34%] Building CXX object CMakeFiles/lightgbm.dir/src/network/linkers_mpi.cpp.o
[ 36%] Building CXX object CMakeFiles/lightgbm.dir/src/network/linkers_socket.cpp.o
[ 38%] Building CXX object CMakeFiles/lightgbm.dir/src/network/network.cpp.o
[ 39%] Building CXX object CMakeFiles/lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o
In file included from /usr/local/src/lightgbm/LightGBM/compute/include/boost/compute/config.hpp:16:0,
                 from /usr/local/src/lightgbm/LightGBM/compute/include/boost/compute/buffer.hpp:14,
                 from /usr/local/src/lightgbm/LightGBM/compute/include/boost/compute/core.hpp:18,
                 from /usr/local/src/lightgbm/LightGBM/src/treelearner/gpu_tree_learner.h:33,
                 from /usr/local/src/lightgbm/LightGBM/src/treelearner/parallel_tree_learner.h:15,
                 from /usr/local/src/lightgbm/LightGBM/src/treelearner/data_parallel_tree_learner.cpp:9:
/usr/local/src/lightgbm/LightGBM/compute/include/boost/compute/cl.hpp:19:10: fatal error: CL/cl.h: No such file or directory
 #include <CL/cl.h>
          ^~~~~~~~~
compilation terminated.
CMakeFiles/lightgbm.dir/build.make:638: recipe for target 'CMakeFiles/lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o' failed
make[2]: *** [CMakeFiles/lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/lightgbm.dir/all' failed
make[1]: *** [CMakeFiles/lightgbm.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by StrikerRUS** [comment]
Hi @Garrus990 !

Can you please provide the full content of your docker file, not only one line?

**Comment by Garrus990** [comment]
Hi @StrikerRUS,
this is the Dockerfile that results in the error message above.

```
ARG cuda_version=10.2
ARG cudnn_version=7
FROM nvidia/cuda:${cuda_version}-cudnn${cudnn_version}-runtime-ubuntu18.04

# prevents python from creation of .pyc files
ENV PYTHONDONTWRITEBYTECODE=true

# Install system packages
RUN apt-get update && apt-get install -y --no-install-recommends --fix-missing \
      bsdmainutils \
      build-essential \
      bzip2 \
      ca-certificates \
      cmake \
      curl \
      htop \
      g++ \
      gcc \
      git \
      graphviz \
      libfreetype6-dev \
      libgl1-mesa-glx \
      libhdf5-dev \
      libpng-dev \
      libzmq3-dev \
      libglib2.0-0 \
      libxext6 \
      libsm6 \
      libxrender1 \
      libboost-dev \
      libboost-system-dev \
      libboost-filesystem-dev \
      openmpi-bin \
      ssh \
      software-properties-common \
      unzip \
      vim \
      wget \
      zsh && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

COPY .bashrc .condarc /root/

# Install conda
ENV CONDA_DIR /opt/conda
ENV PATH $CONDA_DIR/bin:$PATH

RUN wget --quiet --no-check-certificate https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh && \
    /bin/bash /Miniconda3-latest-Linux-x86_64.sh -f -b -p $CONDA_DIR && \
    rm Miniconda3-latest-Linux-x86_64.sh && \
    echo export PATH=$CONDA_DIR/bin:'$PATH' > /etc/profile.d/conda.sh

RUN conda install \
    autopep8 \
    bokeh \
    biopython \
    cudatoolkit=${cuda_version} \
    datashader \
    eli5 \
    hvplot \
    jupyter \
    jupyter_contrib_nbextensions \
    jupyterlab \
    matplotlib \
    mkl \
    mlxtend \
    nbstripout \
    "nodejs>=10.0" \
    numpy \
    pandas \
    plotnine \
    psutil \
    psycopg2 \
    pytables \
    pyyaml \
    scikit-learn \
    scikit-optimize \
    scipy \
    seaborn \
    sqlalchemy \
    tensorflow \
    tqdm \
    wheel \
    xlrd \
    bioconda::pysam && \
    conda clean -yfa

RUN jupyter nbextensions_configurator enable \
    && jupyter contrib nbextension install --user \
    && jupyter nbextension enable execute_time/ExecuteTime \
    && jupyter nbextension enable toc2/main \
    && jupyter labextension install @jupyter-widgets/jupyterlab-manager \
    && jupyter labextension install @pyviz/jupyterlab_pyviz \
    && jupyter labextension install @bokeh/jupyter_bokeh

# Jupyter
EXPOSE 8888

# Add OpenCL ICD files for LightGBM
RUN mkdir -p /etc/OpenCL/vendors && \
    echo "libnvidia-opencl.so.1" > /etc/OpenCL/vendors/nvidia.icd

RUN cd /usr/local/src && mkdir lightgbm && cd lightgbm && \
    git clone --recursive --branch stable --depth 1 https://github.com/microsoft/LightGBM && \
    cd LightGBM && mkdir build && cd build && \
    cmake -DUSE_GPU=1 -DOpenCL_LIBRARY=/usr/local/cuda/lib64/libOpenCL.so -DOpenCL_INCLUDE_DIR=/usr/local/cuda/include/ .. && \
    make OPENCL_HEADERS=/usr/local/cuda-10.2/targets/x86_64-linux/include LIBOPENCL=/usr/local/cuda-10.2/targets/x86_64-linux/lib

ENV PATH /usr/local/src/lightgbm/LightGBM:${PATH}
RUN cd /usr/local/src/lightgbm/LightGBM/python-package && \
   python setup.py install --precompile
RUN apt-get autoremove -y && apt-get clean && \
   rm -rf /var/lib/apt/lists/* && \
   conda clean -a -y

WORKDIR /root

CMD jupyter lab --port=8888 --ip=0.0.0.0 --allow-root --no-browser --NotebookApp.token=''
```

**Comment by StrikerRUS** [comment]
Thanks for your response!

Are these paths OK? 
```
-DOpenCL_LIBRARY=/usr/local/cuda/lib64/libOpenCL.so -DOpenCL_INCLUDE_DIR=/usr/local/cuda/include/
```

Also, try to add the following lines from the original docker:
https://github.com/microsoft/LightGBM/blob/5e0958fbc6071ce4bb8cdc1a5d5526305a4f89b9/docker/gpu/dockerfile.gpu#L15-L17

**Comment by Garrus990** [comment]
Adding the lines with environmental variables did not help - the same error occurs. 

The culprits are, however, the two lines that you suspect - neither the path to `DOpenCL_LIBRARY`, nor to `DOpenCL_INCLUDE_DIR` exist in the folder tree of CUDA 10.2. The file `libOpenCL.so` cannot be found in any subfolder in the `/usr/local` path. Folder named `include` is present only under: `/usr/local/cuda-10.2/targets/x86_64-linux/include`.

**Comment by StrikerRUS** [comment]
Can you please try to search for `libOpenCL.so` across the filesystem?

**Comment by Garrus990** [comment]
I've - nothing.

**Comment by StrikerRUS** [comment]
I guess the reason is in that you are branching from 
```
FROM nvidia/cuda:${cuda_version}-cudnn${cudnn_version}-runtime-ubuntu18.04
```
while our docker is inherited from 
https://github.com/microsoft/LightGBM/blob/18c706dc840c72341a2a85e05dfb5fad6366902a/docker/gpu/dockerfile.gpu#L1

(note `runtime` vs `devel` part). I think it results in different software and drivers installed.

Try with `devel` image, e.g. `nvidia/cuda:10.1-cudnn7-devel-ubuntu18.04`.

**Comment by mirekphd** [comment]
> `runtime` vs `devel` part). I think it results in different software and drivers installed.
> 
> Try with `devel` image, e.g. `nvidia/cuda:10.1-cudnn7-devel-ubuntu18.04`.

I confirm that problem of large `devel` vs. small(er) but insufficient `runtime` base image exists also for CUDA 11: OpenCL dynamic library required by LightGBM is not available in the smallest `nvidia/cuda` images such as `base` or `runtime`, i.e. smaller than `devel`. However,  you can save around 5 GiB of storage space by skipping the cuDNN versions, which contain the `cudnn` token, unless you do deep learning in the same containers as boosted trees.

The smaller `runtime` images - no OpenCL binaries:
```
$ docker run --rm -it nvidia/cuda:11.0-runtime-ubuntu18.04 sh -c "cd / && find | grep CL"
./usr/lib/x86_64-linux-gnu/perl-base/unicore/lib/SB/CL.pl
./usr/lib/x86_64-linux-gnu/perl-base/unicore/lib/Lb/CL.pl
./usr/local/cuda-11.0/targets/x86_64-linux/include/nvtx3/nvtxDetail/nvtxImplOpenCL_v3.h
./usr/local/cuda-11.0/targets/x86_64-linux/include/nvtx3/nvToolsExtOpenCL.h
```

The larger `devel` images, where OpenCL binaries are available:
```
$ docker run --rm -it nvidia/cuda:11.0-devel-ubuntu18.04 sh -c "cd / && find | grep CL.so"
./usr/local/cuda-11.0/targets/x86_64-linux/lib/libOpenCL.so
./usr/local/cuda-11.0/targets/x86_64-linux/lib/libOpenCL.so.1.0.0
./usr/local/cuda-11.0/targets/x86_64-linux/lib/libOpenCL.so.1.0
./usr/local/cuda-11.0/targets/x86_64-linux/lib/libOpenCL.so.1
```

**Comment by mirekphd** [comment]
But there was another twist in the story: `devel` image version is needed after all when CUDA (not just OpenCL as in current lightgbm) is used by a modeling library. This is because only `devel` and larget `nvidia/cuda` images contain the required NVIDIA CUDA compiler (`nvcc`) - at least judging from xgboost requirements, where `cmake` it raises this error in `runtime' images: 
`  Could not find nvcc, please set CUDAToolkit_ROOT.`

Just like for OpenCL, also in this case installation from Ubuntu system repos is possible: the file required by xgboost during compilation is in the `nvidia-cuda-toolkit` package (see [file list](https://packages.ubuntu.com/bionic/amd64/nvidia-cuda-toolkit/filelist)):
`apt install nvidia-cuda-toolkit`

However, as usual, the packages in Ubuntu repos are obsolete - currently by two major CUDA versions, so using `devel` docker image from NVIDIA is the only method to pin the desired CUDA version (here: at 11 not 9.1):
```
The following NEW packages will be installed:
  adwaita-icon-theme [..] g++-6 gcc-6 gcc-6-base [..] libaccinj64-9.1 [..] libcublas9.1 libcudart9.1 libcufft9.1 libcufftw9.1 libcuinj64-9.1 libcups2 libcurand9.1 libcusolver9.1 libcusparse9.1 libnppc9.1 libnppial9.1 libnppicc9.1 libnppicom9.1 libnppidei9.1 libnppif9.1 libnppig9.1 libnppim9.1 libnppist9.1 libnppisu9.1 libnppitc9.1 libnpps9.1 [..] nvidia-cuda-dev nvidia-cuda-doc nvidia-cuda-gdb nvidia-cuda-toolkit
```

**Comment by rsdmse** [comment]
I was directed to this issue from #3286. I confirm that using a multistage build with `devel` and adding the missing libraries to `base` resulted in a working binary that passed the tutorial test (`runtime` also works but it wasn't necessary at least for the tutorial). See #3286 for details.
